fix: Switching accounts in game breaks configs

ryanzhoudev opened the pull request:
This appears to work for now. I only tested once though. If someone else experiences problems let me know.

On hold for new mixin version that allows us to "watch" a variable.

kristofbolyai committed:
chore: Port to 1.21, using NeoForge EventBus (#2597)

* port: Change version to 1.21

* port: Use Java 21 in Github CI

* port: partialTicks -> deltaTracker changes

* port: Misc mixin fixes

* port: ResourceLocation ctor -> factory methods

* port: ClientCommandManager remove deprecated error

* port: BufferBuilder -> ByteBufferBuilder

* port: Port RenderUtils and BufferedRenderUtils (VertexBuilder changes)

* port: Port FontRenderer, MapRenderer and LootrunRenderer (VertexBuilder changes)

* port: ItemStack#getTooltipLines related changes

* port: Port BeaconModel (Entity -> LivingEntity)

* port: Port EntityDataAccessor users (getId -> id)

* port: Port ServerboundUseItemPacket usages

* port: Port WynncraftButtonFeature (fix ConnectScreen#startConnecting arguments, pingServer arguments)

* port: Port BeaconBeamFeature (use Minecraft's BeaconRenderer, remove custom renderer)

* port: Port to using DataComponents

* port: Port AccessWidener changes

* port: Port ClientCommandManager (CommandBuildContext changes)

* port: Port ScoreboardHandler (PlayerScores/ScoreHolder changes)

* port: Port ResourcePackService

* port: Port to 1.21 NeoForge (buildscript, mixins)

This commit ports to 1.21 NeoForge. Notable changes are the fixed NeoForge/Fabric only mixins (I've separated old, require = 0 mixins to Fabric, so it's easier to see what works where). I've also added improvements to our buildscript for both Fabric/NeoForge, to aid version changes.

This commit does not compile, as NeoForge's new EventBus needs some changes to the code base.

* port: Port to NeoForge EventBus

This PR is mostly just import changes. A few rules were added to NeoForge's event bus, to make it easier and less error-prone to use.
Affected parts:
- @Cancelable annotation is now an interface.
- With the annotations removed, the event bus does not need pre-processing, so we no longer depend on EventBusTransformer.
- Generic events were removed. Only PackedEvent used it, it's a simple port.
- Abstract classes cannot register events with @SubscribeEvent. For overlays, I've introduced manager level ticking. For special overlays, I've added WynntilsMod#registerListener to specially register these events.
- Abstract events cannot be listened to. This resulted in a small amount of code duplication. Not enough to consider doing any refactors.

* port: Port CI, README.md to NeoForge

* port: Port RenderUtils#renderCustomNametag (yes, really.)

Between 1.20.2 and 1.21, they removed the - sign from the x coordinate, and it works. Don't ask me.

* fix: Patch modmenu compatibility

* chore: Bump version to 2.0.0-SNAPSHOT

* fix: Fix LootrunRenderer rendering (LevelRenderer PoseStack changes)

* fix: Fix failing tests due to Minecraft's Style class changes

ryanzhoudev committed:
chore: Update mod incompatibilities [skip ci] (#2620)

* chore: Update mod incompatibilities

* remove ias for now

ShadowCat117 committed:
fix: Fix tracked bars not updating (#2622)

ryanzhoudev committed:
fix: Switching accounts in game breaks configs

ShadowCat117 commented:
Should this really be a feature? I feel like it should just be part of the manager.

Also don't forget to reload the storage file too

ryanzhoudev commented:
> Should this really be a feature? I feel like it should just be part of the manager.

Yes, since some people mentioned wanting to be able to use one account's config for another.

> Also don't forget to reload the storage file too

Oh yeah thanks.

ShadowCat117 commented:
> Yes, since some people mentioned wanting to be able to use one account's config for another.

I don't think this solution will fully support that. If you log in as account A, then swap to account B with this feature disabled you will still be saving changes to account A's config. So if you then loaded up as account B first you wouldn't have the changes saved to account A's config.

I think it would just be better to have the global config that was mentioned on discord the other day, but I don't know if/when we plan on doing that

kristofbolyai commented:
Fully agree with ShadowCat. Let's just fix config wipes, nothing else should be done here and now (without an explicit config rework, which won't be a priority during beta).

kristofbolyai committed:
fix: Fix being able to click territories that are not rendered (TerritoryManagementScreen) [skip ci] (#2626)

fix: Fix being able to click territories that are not rendered (TerritoryManagementScreen)

kristofbolyai committed:
fix: Patch SpellCastVignetteFeature until proper action bar parsing [skip ci] (#2630)

fix: Patch SpellCastVignetteFeature until proper action bar parsing

kristofbolyai committed:
fix: Fix world state not being set correctly, fix mana/health bars displaying without data [skip ci] (#2631)

fix: Fix world state not being set correctly, fix mana/health bars displaying without data

kristofbolyai committed:
fix: Fix 1.21 content book parsing [skip ci] (#2629)

fix: Fix 1.21 content book parsing

kristofbolyai committed:
fix: Fix resource pack preloading [skip ci] (#2632)

fix: Fix resource pack preloading

kristofbolyai committed:
fix: Fix failing StyledText tests [skip ci] (#2634)

fix: Fix failing StyledText tests

kristofbolyai committed:
fix: Fix compass menu parsing (#2633)

kristofbolyai committed:
chore: Support fonts in StyledText [skip ci] (#2635)

chore: Support fonts in StyledText

ShadowCat117 committed:
fix: Update containers for 2.1 (#2627)

* wip: Update containers for 2.1

* fix: Fix search bounds

* chore: Shift search widgets

* chore: Add comment

* chore: Disable custom quick jumps

* chore: Shift prof highlight button

* chore: Remove bank name edit button

ryanzhoudev committed:
move to service, storages

ryanzhoudev committed:
Merge branch 'development' into accsw

Commit:
ci: spotless formatting

magicus commented on common/src/main/java/com/wynntils/core/persisted/config/ConfigManager.java:
```diff
@@ -333,4 +333,10 @@ private <T> void setConfigValue(Config<T> config, Object value) {
         }
     }
+
+    public void userSwitched() {
```
I think a better name would be like `reloadAllConfigs` or something like that. The method should describe what it does, not when it is called.

ShadowCat117 committed:
fix: Fix skill point annotator [skip ci] (#2638)

* fix: Fix skill point annotator

* ci: spotless formatting

---------

Co-authored-by: ShadowCat117 <ShadowCat117@users.noreply.github.com>

ShadowCat117 committed:
fix: Fix custom seaskipper screen [skip ci] (#2639)

ShadowCat117 committed:
fix: Fix emerald pouch annotator [skip ci] (#2640)

* fix: Fix emerald pouch annotator

* fix: Check for whitespace too

ryanzhoudev committed:
rename

ryanzhoudev committed:
Merge branch 'accsw' of https://github.com/Wynntils/Artemis into accsw

kristofbolyai committed:
Merge branch 'development' into accsw

kristofbolyai commented:
I have an idea, not sure how feasible. What if you mixin into the Minecraft code, where the user uuid is stored, and basically "inject on change". Iirc there is a way to do that with mixin. If that's possible, you can, well, not signal an event, because it might be too early :), but check for `WynntilsMod#isInitalized`, then call Config/Storage manager directly, to reload the file. 

This is probably a long shot, but I feel like this is the most "complete" way. This also means, you don't have to add a new, kind of weird, feature or service.

ryanzhoudev committed:
fix: Horse keybind no longer relies on soul point item [skip ci] (#2641)

* fix: Horse keybind no longer relies on soul point item

* enable weaps

kristofbolyai committed:
chore: remove (almost) everything related to soul points [skip ci] (#2643)

* chore: remove (almost) everything related to soul points

I've got side tracked while working on the action bar, so I might as well do this now.

* fixup! chore: remove (almost) everything related to soul points

I've got side tracked while working on the action bar, so I might as well do this now.

ryanzhoudev committed:
fix: Teleport scroll annotator beta port [skip ci] (#2642)

fix: Teleport scroll annotator beta port

ShadowCat117 committed:
feat: Rework custom bank page names and quick jumps for 2.1 (#2637)

* chore: bank -> accountBank

* feat: Update quick jumps for 2.1 in personal storage utilities feature

* feat: Update custom page names

* chore: Move emerald count texture when bank panel is visible

* fix: Fix widget not being removed, slight change to input widget

* refactor: Consts

---------

Co-authored-by: Kristof <49001742+kristofbolyai@users.noreply.github.com>

kristofbolyai committed:
feat: Support the 2.1 action bar (custom bars, disabling parts) (#2636)

* feat: Support the 2.1 action bar (custom bars, disabling parts)

* feat: Stable-ish parsing for health/mana bars and texts

* feat: Parsed all action bar segment types

* feat: Move matchers to their respective models, finalize spell parsing

* feat: finalize

* chore: touchups

ShadowCat117 committed:
feat: Add ability tree reset and aspects containers [skip ci] (#2645)

ShadowCat117 committed:
chore: Update bank textures [skip ci] (#2644)

ShadowCat117 committed:
fix: Fix crashing in character selection screen (#2646)

kristofbolyai committed:
fix: Fix health bar parsing when health is "critical" [skip ci] (#2647)

fix: Fix health bar parsing when health is "critical"

kristofbolyai committed:
fix: Fix MaterialAnnotator crashing when trying to look up "Unprocessed" items [skip ci] (#2648)

fix: Fix MaterialAnnotator crashing when trying to look up "Unprocessed" items

ryanzhoudev commented:
> I have an idea, not sure how feasible. What if you mixin into the Minecraft code, where the user uuid is stored, and basically "inject on change". Iirc there is a way to do that with mixin. If that's possible, you can, well, not signal an event, because it might be too early :), but check for `WynntilsMod#isInitalized`, then call Config/Storage manager directly, to reload the file.
> 
> This is probably a long shot, but I feel like this is the most "complete" way. This also means, you don't have to add a new, kind of weird, feature or service.

I'm not sure this will work. The `User user` variable is `private final`, and IAS uses an accessor to mess with it. But it's also possible that some other mods or clients use AW, which if we inject into the getUser method, we won't detect. I didn't find anything that allows us to "watch" a variable. Thoughts?

ryanzhoudev committed:
Merge branch 'development' into accsw

ryanzhoudev commented:
Not interested in trying to update and resolve this mess, when the new mixins stuff comes out we can look at it again. I will make an issue.
